cranelift/codegen/src/isa/zkasm/lower.isle

```diff
@@ -77,7 +77,7 @@
   (zk_mul x y))
 
 (rule 1 (lower (has_type (fits_in_32 (ty_int ty)) (imul x y)))
-  (zk_mul x y))
+  (zk_mul_32 x y))
```

Comment:
How about we try to use the super-power of ISLE here and instead of introducing another intermediate rule, directly lower this to `zk_mul` plus `and`:

```
(rule 1 (lower (has_type (fits_in_32 (ty_int ty)) (imul x y)))
  (gen_andi (zk_mul x y) 0xFFFFFFFF))
```

Reply:
I think it would be better make ISLE improvements in another PR, since this one is quite big already and such changes aren't necessary for first working version of 32-bit computations

Reply:
Sounds good - let's file an issue to investigate pros and cons between these two approaches and work on this in the future PRs.